file.c

```diff
@@ -3596,34 +3631,298 @@ rb_home_dir_of(VALUE user, VALUE result)
 }
 
 #ifndef _WIN32
+VALUE rb_getlogin(void);
+VALUE rb_getpwdirnam_for_login(void);  /* read as: "get pwd db home dir by username for login" */
+VALUE rb_getpwdiruid(void);            /* read as: "get pwd db home dir by uid" */
+
 VALUE
 rb_default_home_dir(VALUE result)
 {
     const char *dir = getenv("HOME");
+    if (dir)
+        return copy_home_path(result, dir);
 
-#if defined HAVE_PWD_H
-    if (!dir) {
-	const char *login = getlogin();
-	if (login) {
-	    struct passwd *pw = getpwnam(login);
-	    if (pw) {
-		copy_home_path(result, pw->pw_dir);
-		endpwent();
-		return result;
-	    }
-	    endpwent();
-	    rb_raise(rb_eArgError, "couldn't find HOME for login `%s' -- expanding `~'",
-		     login);
-	}
-	else {
-	    rb_raise(rb_eArgError, "couldn't find login name -- expanding `~'");
-	}
+    /* We'll look up the user's default home dir in the password db by login
+     * name, if possible, and failing that will fall back to looking the
+     * information up by uid (as would be needed for processes that are not a
+     * descendant of login(1) or a work-alike).
+     *
+     * While the lookup by uid is more likely to succeed (since we always have
+     * a uid, but may or may not have a login name), we prefer first looking
+     * up by name to accommodate the possibility of multiple login names (each
+     * with its own record in the password database, so each with a
+     * potentially different home directory) being mapped to the same uid (as
+     * explicitly allowed for by POSIX; see getlogin(3posix)).
+     */
+
+    VALUE pw_dir_str_nm = rb_getpwdirnam_for_login();
+    if (!NIL_P(pw_dir_str_nm)) {
+        /* found it */
+        copy_home_path(result, RSTRING_PTR(pw_dir_str_nm));
+        rb_str_resize(pw_dir_str_nm, 0);
+        return result;
+    }
+
+    VALUE pw_dir_str_uid = rb_getpwdiruid();
+    if (!NIL_P(pw_dir_str_uid)) {
+        /* found it */
+        copy_home_path(result, RSTRING_PTR(pw_dir_str_uid));
+        rb_str_resize(pw_dir_str_uid, 0);
+        return result;
+    }
+
+    rb_raise(rb_eArgError, "couldn't find home for uid `%ld'", (long)getpid());
```

Comment:
Preferably, the error messages should be kept for each case, for the backward compatibilities.
And some compilers don't recognize `NORETURN` function but warn that no return at the end.
`return` for `getenv("HOME")` is placed there for that reason.

Reply:
> Preferably, the error messages should be kept for each case, for the backward compatibilities.

With the code always falling back on doing the lookup by uid, the case for the first error message will never happen:
```ruby
           rb_raise(rb_eArgError, "couldn't find HOME for login `%s' -- expanding `~'",
                    login);
```

But I'll put the bottoming-out error message back to:
```
           rb_raise(rb_eArgError, "couldn't find login name -- expanding `~'");
```
I had changed that because the specific "`...expanding `~'...`" wording was a bit of a misnomer, and with the lookup by uid even more so. But I see your point about backward compatibility in this regard.


> And some compilers don't recognize `NORETURN` function but warn that no return at the end.
> `return` for `getenv("HOME")` is placed there for that reason.

Ah, I see. I'll have to account for that type of thing in a couple of places with my changes.

Thanks for all the great feedback, @nobu. I'll incorporate these changes, and will also look to see about whether pieces of them might live more comfortably on in `process.c`, [as you mentioned in the issue tracker](https://bugs.ruby-lang.org/issues/16787?issue_count=2&issue_position=1&next_issue_id=12226#note-11).